### ISSUE TITLE
Model Reproducibility 

### ISSUE THREAD

**Issue by Alexisbiancotti** [MaartenGr/BERTopic#987]
Hello all,

I am encountering issues with reproducibility of the model. Every time I run it, the resulting topics vary greatly. I have tried setting random values as per suggestions found on various online forums, but they not work. Could anyone help me figure out what I am doing wrong? Here is my code:

np.random.seed(42)

specialWords = ["support","security","maintenance","protection","subscription"]

umap_model = UMAP(n_neighbors=10, 
                  n_components=3, 
                  min_dist=0.5,
                  target_weight=0.5,
                  metric='cosine', 
                  random_state=100)

topic_model = BERTopic(umap_model=umap_model,
                       language="english",
                       seed_topic_list = specialWords,
                       top_n_words=2,
                       nr_topics=50)

topics, probabilities = topic_model.fit_transform(df['descp_imprv'])

Notes: I try with jupyter notebook and spyder and both have the same issue

Thanks in advance!

**Comment by MaartenGr** [comment]
Which version of BERTopic are you using? Also, are you using the GPU-accelerated version of UMAP? Lastly, could you try it out in a completely fresh environment, perhaps something went wrong with the installation.

**Comment by Alexisbiancotti** [comment]
Hi, i leave you the answers:
1) The version is "0.13.0"
2) I don't use that version of umap (I believe), i import it with "from umap import UMAP"
3)  I try a creating a new environment and the error is the same (i use 35 k texts to train it)

**Comment by MaartenGr** [comment]
For me, the following is working correctly and gives the exact same topics twice:

```python
from umap import UMAP
from bertopic import BERTopic
import numpy as np

np.random.seed(42)

specialWords = [["support","security","maintenance","protection","subscription"]]

# Create topics
umap_model = UMAP(n_neighbors=10, n_components=3, min_dist=0.5,
                  target_weight=0.5, metric='cosine', random_state=100)
topic_model = BERTopic(umap_model=umap_model,
                       language="english",
                       seed_topic_list = specialWords,
                       top_n_words=2,
                       nr_topics=50)
topics, probabilities = topic_model.fit_transform(docs)

# Create topics in the same way
umap_model = UMAP(n_neighbors=10, n_components=3, min_dist=0.5,
                  target_weight=0.5, metric='cosine', random_state=100)
topic_model1 = BERTopic(umap_model=umap_model,
                       language="english",
                       seed_topic_list = specialWords,
                       top_n_words=2,
                       nr_topics=50)
topics1, probabilities1 = topic_model1.fit_transform(docs)

# Compare topics
assert topics == topics1
```

For me, the code above makes the model reproducible, so it might be worthwhile to check out your original code to see if it matches exactly with the above. Also, note that I made `specialWords` a list of lists as that is expected in BERTopic.

**Comment by Alexisbiancotti** [comment]
Thanks for the response,

I try the code that you send me and I have the same problem, I made some aditional tests and found that this error is "erratic" dependending on the qty train texts.

My train data have arrond 35k texts, if I use the first 2000 (df[:2000]) the results of each moedel are the same but if i use the first 3000 there are differents.

I made this test in different computers, both having a fresh env with only bertopic (and its dependencies) and erasing the variables after each run,

I saw the training data and I couldnt find something that may trigger this difference, do you have any clue what might be?

**Comment by MaartenGr** [comment]
Could you, instead of using your own docs, use the following docs instead:

```python
from sklearn.datasets import fetch_20newsgroups
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']
```

That way, the following becomes a self-contained example:

```python
from umap import UMAP
from bertopic import BERTopic
import numpy as np
from sklearn.datasets import fetch_20newsgroups
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

np.random.seed(42)

specialWords = [["support","security","maintenance","protection","subscription"]]

# Create topics
umap_model = UMAP(n_neighbors=10, n_components=3, min_dist=0.5,
                  target_weight=0.5, metric='cosine', random_state=100)
topic_model = BERTopic(umap_model=umap_model,
                       language="english",
                       seed_topic_list = specialWords,
                       top_n_words=2,
                       nr_topics=50)
topics, probabilities = topic_model.fit_transform(docs)

# Create topics in the same way
umap_model = UMAP(n_neighbors=10, n_components=3, min_dist=0.5,
                  target_weight=0.5, metric='cosine', random_state=100)
topic_model1 = BERTopic(umap_model=umap_model,
                       language="english",
                       seed_topic_list = specialWords,
                       top_n_words=2,
                       nr_topics=50)
topics1, probabilities1 = topic_model1.fit_transform(docs)

# Compare topics
assert topics == topics1
```

If you run the above, do you then get any assertion errors?

**Comment by Alexisbiancotti** [comment]
Hi,

I runned your example 3 times and it always returned the same results:

![imagen](https://user-images.githubusercontent.com/54298035/217854672-c240b868-9b23-45de-9065-d7199cc84817.png)

To further explain the error that i encounter with my own data, this is the results of the same script with 2000 training texts (no assertion error)
![imagen](https://user-images.githubusercontent.com/54298035/217856941-a77ce7c2-e209-4f50-955d-a9ee727f7882.png)

Then with 3000 (gives the error)
![imagen](https://user-images.githubusercontent.com/54298035/217857295-09ada36a-cdb0-46f4-be3b-4ebc6fcf8f0c.png)

**Comment by MaartenGr** [comment]
My guess is that there is something going on with your dataset. More specifically, perhaps there is some randomness being created in your dataset seeing as I cannot reproduce the issue myself with a different dataset.

**Comment by maltehueckstaedt** [comment]
> Hi,
> 
> I runned your example 3 times and it always returned the same results:
> 
> ![imagen](https://user-images.githubusercontent.com/54298035/217854672-c240b868-9b23-45de-9065-d7199cc84817.png)
> 
> To further explain the error that i encounter with my own data, this is the results of the same script with 2000 training texts (no assertion error) ![imagen](https://user-images.githubusercontent.com/54298035/217856941-a77ce7c2-e209-4f50-955d-a9ee727f7882.png)
> 
> Then with 3000 (gives the error) ![imagen](https://user-images.githubusercontent.com/54298035/217857295-09ada36a-cdb0-46f4-be3b-4ebc6fcf8f0c.png)

@Alexisbiancotti  Have you found a solution to the problem? Unfortunately I have exactly the same problem!

**Comment by Alexisbiancotti** [comment]
Hi, no i Haven´t. Finally I abandoned the idea :(
